File: board.py
from dependencies import *

from move import Move
from ttt import TicTacToe

class SubBoard:

	def __init__(self):
		self.board = np.full((3, 3), None)

	def __repr__(self):
		return str(self.board)

	def __getitem__(self, coord: tuple[int, int]) -> int:
		return self.board[coord]

	@property
	def empty(self) -> set[tuple[int, int]]:
		return {(i, j) for i in range(3) for j in range(3) if self.board[i, j] is None}

	@property
	def winner(self) -> int:
		"""who has won the SubBoard

		Returns:
				int: None if no winner yet; 0 if tie; -1 or 1 if player has won
		"""

		# check rows
		for i in range(3):
			if self.board[i, 0] == self.board[i, 1] == self.board[i, 2] and self.board[i, 0] is not None:
				return self.board[i, 0]

		# check columns
		for j in range(3):
			if self.board[0, j] == self.board[1, j] == self.board[2, j] and self.board[0, j] is not None:
				return self.board[0, j]

		# check diagonals
		if self.board[0, 0] == self.board[1, 1] == self.board[2, 2] and self.board[0, 0] is not None:
			return self.board[0, 0]
		if self.board[0, 2] == self.board[1, 1] == self.board[2, 0] and self.board[0, 2] is not None:
			return self.board[0, 2]

		# check for tie
		if self.empty == set():
			return 0

		# no winner yet
		return None

	@property
	def as_tuple(self):
		return tuple(map(tuple, self.board))

	def validate_move(self, row: int, col: int) -> bool:
		if ((row, col) in self.empty):
			return True

	def move(self, row: int, col: int, iplayer: int) -> bool:
		self.board[row, col] = iplayer

class Board:
	value_map = {-1: 'O',
							 None: '~',
							 1: 'X'}

	lines = {
			((0, 0), (0, 1), (0, 2)), ((0, 2), (1, 1), (2, 0)), ((2, 0), (2, 1), (2, 2)), # rows
			((0, 1), (1, 1), (2, 1)), ((0, 0), (1, 1), (2, 2)), ((1, 0), (1, 1), (1, 2)), # columns
			((0, 0), (1, 0), (2, 0)), ((0, 2), (1, 2), (2, 2)) # diagonals
		}

	with open('ttt.pkl', 'rb') as f:
		fitness = pickle.load(f)

	def __init__(self):
		# big board
		# None if nobody has won the subboard
		# 0 or 1 if player 0 or 1 has won
		self.board = np.full((3, 3), None)
		self.subboards = np.array([deepcopy([deepcopy(SubBoard()) for row in range(3)]) for col in range(3)]) # 9 subboards arranged in 3 x 3 grid

		self.iplayer = 1 # tracks which player should go
		self.previous: Move = None # previous move. None represents no previous moves

		# TEMP: subject to change
		self.importance = np.array([ # importance score of each subboard CHANGE
			[1, 0, 1],
			[0, 2, 0],
			[1, 0, 1],
		], dtype=float)

		self.initial = self.importance.copy()

	def __repr__(self):
		out = ''

		out += '  0 1 2  3 4 5  6 7 8\n'

		for row in range(9):
			if (row % 3 == 0 and row != 0):
				out += '\n'
			out += str(row) + ' '

			for col in range(9):
				move = Move(row=row, col=col)
				(bigrow, bigcol), (subrow, subcol) = move.relative

				if (self.board[bigrow, bigcol] is None):
					value = self.value_map[self.subboards[bigrow, bigcol][subrow, subcol]]
				else:
					value = self.value_map[self.subboards[bigrow, bigcol].winner]

				if (col % 3 == 0 and col != 0):
					out += ' '

				out += value + ' '

			out += '\n'

		return out

	@property
	def absolute(self) -> np.array:
		absolute = np.empty((9, 9), dtype=int)
		for row in range(9):
			for col in range(9):
				move = Move(row=row, col=col)
				(bigrow, bigcol), (subrow, subcol) = move.relative
				absolute[row, col] = self.subboards[bigrow, bigcol][subrow, subcol]

		return absolute

	def next_subboard(self) -> tuple[int, int]:
		"""returns the next subboard to play in; -1 if any

		Returns:
				tuple[int, int]: the next subboard to play in
		"""

		bigrow, bigcol = self.previous.sub

		if (self.subboards[bigrow, bigcol].winner is None):
			return bigrow, bigcol

		return (-1, -1)

	def possible_moves(self) -> list[Move]: # BUG: make sure the subboard is corresponding
		"""finds the possible moves that can be made

		Returns:
				set[Move]: set of possible moves
		"""

		moves = []
		for row in range(9):
			for col in range(9):
				move = Move(row=row, col=col)
				if (self.validate_move(move)):
					moves.append(move)

		return moves

	def validate_move(self, move: Move) -> bool:
		"""validates a move

		Args:
				move (Move): move to validate

		Returns:
				bool: True if the move is valid, False otherwise
		"""

		(bigrow, bigcol), (subrow, subcol) = move.relative

		if not (self.board[bigrow, bigcol] is None): # if the subboard that move is trying to play in has already been won
			return False

		if not ((subrow, subcol) in self.subboards[bigrow, bigcol].empty): # if the move has already been made
			return False

		if (self.previous is not None):
			if not (move.bigrow == self.previous.subrow and move.bigcol == self.previous.subcol): # if it's not corresponding
				if not (self.subboards[self.previous.subrow, self.previous.subcol].winner is None): # if the previous subboard has not been won
					return True
				return False

		return True

	def bwin(self) -> int: # REFACTOR: refactor using self.lines
		"""who has won the board

		Returns:
				int: None if no winner yet; 0 if tie; -1 or 1 if player has won
		"""

		# check rows
		for row in range(3):
			if self.board[row, 0] == self.board[row, 1] == self.board[row, 2] and self.board[row, 0] is not None:
				return self.board[row, 0]

		# check columns
		for col in range(3):
			if self.board[0, col] == self.board[1, col] == self.board[2, col] and self.board[0, col] is not None:
				return self.board[0, col]

		# check diagonals
		if self.board[0, 0] == self.board[1, 1] == self.board[2, 2] and self.board[0, 0] is not None:
			return self.board[0, 0]
		if self.board[0, 2] == self.board[1, 1] == self.board[2, 0] and self.board[0, 2] is not None:
			return self.board[0, 2]

		# check for tie
		for row in self.board:
			for col in row:
				if row is None:
					return None

		# no winner yet
		return None

	def move(self, row: int, col: int) -> bool:
		"""makes a move

		Args:
				move (Move): move to make

		Returns:
				bool: True if the move was made, False otherwise
		"""

		move = Move(row=row, col=col)

		if not (self.validate_move(move)):
			print(move)
			print('Invalid Move')
			print(f'{self.value_map[self.iplayer]} to move')

			return False

		(bigrow, bigcol), (subrow, subcol) = move.relative

		self.subboards[bigrow, bigcol].move(subrow, subcol, self.iplayer)

		status = self.subboards[bigrow, bigcol].winner
		if (status is not None): # if the subboard game has ended
			self.board[bigrow][bigcol] = status

		self.iplayer = -self.iplayer
		self.previous = move

		if ((winner := self.bwin()) is not None):
			if (winner == 1):
				print('X wins')
			elif (winner == -1):
				print('O wins')
			elif (winner == 0):
				print('Tie')

		return True

	# returns the average importance score of all the subboards that have not been won
	def total_importance(self) -> float:
		total = 0
		counter = 0
		for row in range(3):
			for col in range(3):
				if (self.board[row, col] is None):
					total += self.importance[row, col]
					counter += 1

		return total / counter

	def scale_importance(self, lower: float=0, upper: float=10):
		smallest = np.amin(self.importance)
		largest = np.amax(self.importance)

		def f(x):
			m = (upper - lower) / (largest - smallest)

			return m * (x - smallest) + lower

		transformation = np.vectorize(f)

		self.importance = transformation(self.importance)

	def ai_move(self, bdumb=False) -> Move:
		def dumb():
			possible = self.possible_moves()
			for move in possible:
				hypothetical = deepcopy(self)
				row, col = move.absolute
				bigrow, bigcol = move.big
				subrow, subcol = move.sub

				if (self.board[subrow, subcol] is not None):
					continue

				hypothetical.move(row, col)

				if (hypothetical.board[bigrow, bigcol] == self.iplayer):
					return move

			move = random.choice(self.possible_moves())

			return move

		if (bdumb):
			dumb()

		for bigrow in range(3):
			for bigcol in range(3):
				subboard = self.subboards[bigrow, bigcol]

				# self vs. opponent move count

				count = 0
				for subrow in range(3):
					for subcol in range(3):
						# if (self.board[subrow, subcol] is not None):
							# print('Won')
							# count += 5 * self.total_importance()
							# count += 5000000
						if (subboard[subrow, subcol] == -self.iplayer):
							count += 1
						elif (subboard[subrow, subcol] == self.iplayer):
							count -= 1

				self.importance[bigrow, bigcol] += count

		# same as line 358
		# increase importance if the subboard is won
		# for row in range(3):
			# for col in range(3):
				# if (self.board[row, col] is not None):
					# self.importance[row, col] += 50000

		# the subboards that can lead to a win have higher scores

		# raise the score of all subboards in the line by their average importance
		shift = np.full((3, 3), 0, dtype=float)
		for a, b, c in self.lines: # TODO:
			average = (abs(self.importance[a]) + abs(self.importance[b]) + abs(self.importance[c])) / 3

			# if (a != None or b != None or c != None): # CONFUCIOUS: idk why this is here
				# average *= -1

			shift[a] += average
			shift[b] += average
			shift[c] += average

		self.importance += shift

		self.scale_importance()

		# add importance to subboards that have been won
		for row in range(3):
			for col in range(3):
				if (self.board[row, col] is not None):
					self.importance[row, col] = 10

		# possible = [(row, col) for row in range(3) for col in range(3) if subboard[row, col] is None]

		scores = sorted(self.importance.flatten(), reverse=True)

		bigrow, bigcol = self.next_subboard()
		subboard = self.subboards[bigrow, bigcol]

		if ((bigrow, bigcol) == (-1, -1)): # TODO: handle case where the player is sent to a taken subboard
			bigrow, bigcol = random.choice(np.where(self.importance == np.amax(self.importance)))

		if (self.importance[bigrow, bigcol] >= scores[2]): # BUG: sending me to taken square
			print('Important')

			ttt = TicTacToe(self.subboards[bigrow, bigcol], self.iplayer)

			for row in range(3):
				for col in range(3):
					if (self.board[row, col] is not None):
						hypothetical = deepcopy(ttt)
						hypothetical.move(row, col)

						score = self.fitness[hypothetical] # BUG: not all states are in self.fitness ex: two O in one subboard

						self.importance[row, col] -= 2 * score

			scores = sorted(self.importance.flatten())
			for score in scores: # BUG: if all of the important subboards are taken, this will cause an error
				# TODO: use numpy.amax to find the highest score

				useless_coords = list(map(tuple, np.asarray(np.where(np.isclose(self.importance, score))).T))
				useless_coords = [coord for coord in useless_coords if subboard[coord] is None]

				if (useless_coords):
					selection = random.choice(np.transpose(np.where(self.importance == np.amax(self.importance))))
					break

		if (self.importance[bigrow, bigcol] < scores[2]): # if the player was not sent to an important subboard
			# useless = np.asarray(np.where(self.importance == scores[-1])).T
			# selection = random.choice(useless)
			for score in scores[::-1]:
				useless_coords = list(map(tuple, np.asarray(np.where(np.isclose(self.importance, score))).T))
				useless_coords = [coord for coord in useless_coords if subboard[coord] is None]

				if (useless_coords):
					selection = random.choice(useless_coords)
					break

			print('not important')

		# elif (self.importance[bigrow, bigcol] > scores[2]): # if the player was sent to an important subboard
			# TODO: pick the best move that will do the least damange

			# scores = sorted(self.importance.flatten(), reverse=True)

		move = Move(bigrow=bigrow, bigcol=bigcol, subrow=selection[0], subcol=selection[1])

		print(self.importance)

		return move

if __name__ == '__main__':
	board = Board()

	moves = [
		Move(row=0, col=1),
		Move(row=0, col=3),
		Move(row=1, col=1),
		Move(row=3, col=3),
		Move(row=2, col=1),
	]

	# # move = Move(row=0, col=0)
	[board.move(move) for move in moves]

	open('moves.log', 'a').write('\n\n')
	while (True):
		print(board, flush=True)

		row, col = map(int, input().split())
		print()

		move = Move(row=row, col=col)
		board.move(move)

		with open('log', 'a') as f:
			f.write(str(row) + ' ' + str(col) + '\n')

	# move = Move(1, 2, 3, 4)
	# print(type(move))

File: src/javascript/board.js
function deepClone(obj) {
	if (obj === null || typeof obj !== "object")
		return obj
	var props = Object.getOwnPropertyDescriptors(obj)
	for (var prop in props) {
		props[prop].value = deepClone(props[prop].value)
	}
	return Object.create(
		Object.getPrototypeOf(obj),
		props
	)
}

class Move {
	// constructor(row, col, bigrow, bigcol, subrow, subcol, iplayer) {
	constructor(...args) {
		var row = args[0]; var col = args[1]; var bigrow = args[2]; var bigcol = args[3]; var subrow = args[4]; var subcol = args[5]; var iplayer = args[6];

		var absolute;
		if (row == undefined || col == undefined) {
			absolute = false;
		} else if (bigrow == undefined || bigcol == undefined || subrow == undefined || subcol == undefined) {
			absolute = true;
		}

		var bigrow, bigcol, subrow, subcol;
		if (absolute) {
			bigrow = Math.floor(row / 3);
			subrow = row % 3;

			bigcol = Math.floor(col / 3);
			subcol = col % 3;
		} else {
			row = bigrow * 3 + subrow;
			col = bigcol * 3 + subcol;
		}

		this.row = row;
		this.col = col;
		this.bigrow = bigrow;
		this.bigcol = bigcol;
		this.subrow = subrow;
		this.subcol = subcol;

		this.iplayer = iplayer;
	}

	relative() {
		return [[this.bigrow, this.bigcol], [this.subrow, this.subcol]];
	}

	absolute() {
		return [this.row, this.col];
	}

	big() {
		return [this.bigrow, this.bigcol];
	}

	sub() {
		return [this.subrow, this.subcol];
	}

	bcross() {
		return this.bigrow == this.bigcol || this.bigrow + this.bigcol == 2;
	}

	*group(bigrow, bigcol) {
		var moves = [];
		for (let subrow = 0; subrow < 3; subrow++) {
			for (let subcol = 0; subcol < 3; subcol++) {
				yield new Move(null, null, bigrow, bigcol, subrow, subcol, null);
				// moves.push(Move(bigrow = this.bigrow, bigcol = this.bigcol, subrow = subrow, subcol = subcol));
			}
		}

		return moves;
	}

	id() {
		let [row, col] = this.absolute();
		return row + ' ' + col;
	}
}

class SubBoard {
	constructor() {
		this.board = [
			[null, null, null],
			[null, null, null],
			[null, null, null],
		];

		this.score = 0;
	}

	get(row, col) {
		return this.board[row][col];
	}

	empty() {
		var coords = [];
		for (let row = 0; row < 3; row++) {
			for (let col = 0; col < 3; col++) {
				if (this.board[row][col] == null) {
					coords.push([row, col]);
				}
			}
		}

		return coords;
	}

	winner() {
		for (let i = 0; i < 3; i++) {
			if (this.board[i][0] == this.board[i][1] && this.board[i][1] == this.board[i][2] && this.board[i][0] != null) {
				return this.board[i][0];
			}
		}

		for (let j = 0; j < 3; j++) {
			if (this.board[0][j] == this.board[1][j] && this.board[1][j] == this.board[2][j] && this.board[0][j] != null) {
				return this.board[0][j];
			}
		}

		if (this.board[0][0] == this.board[1][1] && this.board[1][1] == this.board[2][2] && this.board[0][0] != null) {
			return this.board[0][0];
		}
		if (this.board[0][2] == this.board[1][1] && this.board[1][1] == this.board[2][0] && this.board[0][2] != null) {
			return this.board[0][2];
		}

		if (this.empty().length == 0) {
			return 0;
		}

		return null;
	}

	validate_move(row, col) {
		if (this.empty().includes([row, col])) {
			return true;
		}

		return false;
	}

	next_subboard() {
		var [bigrow, bigcol] = this.previous.sub();

		if (this.subboards[bigrow][bigcol].winner() == null) {
			return [bigrow, bigcol];
		}

		return [-1, -1];
	}

	move(row, col, iplayer) {
		this.board[row][col] = iplayer;
	}

	undo(row, col) {
		this.board[row][col] = null;
	}
}

class Board {
	constructor(aiPlayer) {
		this.value_map = {
			'-1': 'O',
			null: '~',
			1: 'X',
		};

		this.board = [
			[null, null, null],
			[null, null, null],
			[null, null, null],
		];

		this.subboards = [
			[new SubBoard(), new SubBoard(), new SubBoard()],
			[new SubBoard(), new SubBoard(), new SubBoard()],
			[new SubBoard(), new SubBoard(), new SubBoard()],
		];

		this.lines = [
			[[0, 0], [0, 1], [0, 2]], [[0, 2], [1, 1], [2, 0]], [[2, 0], [2, 1], [2, 2]],
			[[0, 1], [1, 1], [2, 1]], [[0, 0], [1, 1], [2, 2]], [[1, 0], [1, 1], [1, 2]],
			[[0, 0], [1, 0], [2, 0]], [[0, 2], [1, 2], [2, 2]],
		];

		this.iplayer = 1;
		this.previous = null;

		this.aiPlayer = aiPlayer;
		this.score = 0;
	}

	__repr__() {
		var out = '';
		out += '  0 1 2  3 4 5  6 7 8\n';

		var move;
		var bigrow, bigcol, subrow, subcol;
		var value;

		for (let row = 0; row < 9; row++) {
			if (row % 3 == 0 && row != 0) {
				out += '\n';
			}
			out += row + ' ';

			for (let col = 0; col < 9; col++) {
				move = new Move(row = row, col = col);
				bigrow = move.bigrow; bigcol = move.bigcol; subrow = move.subrow; subcol = move.subcol;

				if (this.board[bigrow][bigcol] == null) {
					value = this.value_map[this.subboards[bigrow][bigcol].get(subrow, subcol)];
				} else {
					value = this.value_map[this.subboards[bigrow][bigcol].winner()];
				}

				if (col % 3 == 0 && col != 0) {
					out += ' ';
				}

				out += value + ' ';
			}

			out += '\n';
		}

		return out;
	}

	next_subboard() {
		var [bigrow, bigcol] = this.previous.sub();

		if (this.subboards[bigrow][bigcol].winner() == null) {
			return [bigrow, bigcol];
		}

		return [-1, -1];
	}

	*possible_moves() {
		var moves = [];

		var move;
		for (let row = 0; row < 3; row++) {
			for (let col = 0; col < 3; col++) {
				move = new Move(row = row, col = col);

				if (this.validate_move(move)) {
					// moves.push(move);
					yield move;
				}
			}
		}

		return moves;
	}

	contains(coords, target) {
		for (let i = 0; i < coords.length; i++) {
			if (coords[i][0] == target[0] && coords[i][1] == target[1]) {
				return true;
			}
		}

		return false;
	}

	validate_move(move) {
		var relative = move.relative();
		var [bigrow, bigcol] = relative[0];
		var [subrow, subcol] = relative[1];

		if (this.board[bigrow][bigcol] != null) {
			return false;
		}

		var empty = this.subboards[bigrow][bigcol].empty()
		if (!this.contains(empty, [subrow, subcol])) {
			return false;
		}

		if (this.previous != null) {
			if (move.bigrow != this.previous.subrow || move.bigcol != this.previous.subcol) {
				if (this.subboards[this.previous.subrow][this.previous.subcol].winner() != null) {
					return true;
				}
				return false;
			}
		}

		return true;
	}

	bwin() {
		for (let row = 0; row < 3; row++) {
			if (this.board[row][0] == this.board[row][1] == this.board[row][2] && this.board[row][0] != null) {
				return this.board[row][0];
			}
		}

		for (let col = 0; col < 3; col++) {
			if (this.board[0][col] == this.board[1][col] == this.board[2][col] && this.board[0][col] != null) {
				return this.board[0][col];
			}
		}

		if (this.board[0][0] == this.board[1][1] == this.board[2][2] && this.board[0][0] != null) {
			return this.board[0][0];
		}
		if (this.board[0][2] == this.board[1][1] == this.board[2][0] && this.board[0][2] != null) {
			return this.board[0][2];
		}

		for (let i = 0; i < 3; i++) {
			for (let j = 0; j < 3; j++) {
				if (this.board[i][j] == null) {
					return null;
				}
			}
		}
		return null;
	}

	move(row, col) {
		var move = new Move(row = row, col = col);

		if (!this.validate_move(move)) {
			console.log('Invalid move');
			return false;
		}

		var relative = move.relative();

		var [bigrow, bigcol] = relative[0];
		var [subrow, subcol] = relative[1];

		this.subboards[bigrow][bigcol].move(subrow, subcol, this.iplayer);

		var state = this.subboards[bigrow][bigcol].winner();
		if (state != null) {
			this.board[bigrow][bigcol] = state;
		}

		this.iplayer = -this.iplayer;
		this.previous = move;

		var winner = this.bwin();
		if (winner != null) {
			switch (winner) {
				case 1:
					console.log('X wins!');
					break;
				case -1:
					console.log('O wins!');
					break;
				case 0:
					console.log('Draw!');
					break;
			}
		}

		return true;
	}

	undo() {
		this.iplayer *= -1;

		var relative = this.previous.relative();
		var [bigrow, bigcol] = relative[0];
		var [subrow, subcol] = relative[1];

		this.subboards[bigrow][bigcol].undo(subrow, subcol);
		this.board[bigrow][bigcol] = null;
	}

	minimax(depth, player, alpha, beta) {
		var moves = Array.from(this.possible_moves());

		var score, best_move;

		if (depth == 0 || moves.length == 0) {
			score = this.getScore();

			return {
				score: score,
				move: null,
			}
		}

		for (let move of moves) {
			this.move(move);
			score = this.minimax(depth + 1, player == 1, alpha, beta).score;

			if (player == this.aiPlayer) {
				if (score > alpha) {
					alpha = score;
					best_move = move;
				}
			} else {
				if (score < beta) {
					beta = score;
					best_move = move;
				}
			}

			this.undo();
			if (alpha > beta) {
				break;
			}
		}

		return {
			score: player == this.aiPlayer ? alpha : beta,
			move: best_move,
		};
	}

	// getScore() {

	// }

	// scoreLine(line) {
		// var counter = {
			// X: 0,
			// O: 0,
			// tie: 0,
		// };

		// for ()
	// }
}